### PR TITLE
Ensure the sempahore is released

### PR DESCRIPTION
**Pull request by spicavigo**
1. Minor bug in L120 (let_all_through was under another if condition)
2. Add finally to all _let_one_through stuff
3. Modified the let_all_through logic to just release your own lock instead of releasing multiple times 

@CurleySamuel @hzhaofb @dcreemer

### CODE REVIEW

File: exceptions.py [hunk 1]
```diff
@@ -243,12 +249,12 @@ def _let_one_through(exception, data):
     if my_sem.acquire(blocking=False):
         # Look at me - I'm the captain now.
         return True
-        pass
     else:
         # Someone else is already handling
         # the exception. Sit here until they
         # release the semaphore.
         my_sem.acquire()
+        my_sem.release()
```
Comment (hzhaofb) [inline exceptions.py:257]: I don't fully understand this change. Please explain.

Comment (spicavigo) [inline exceptions.py:257]: The previous logic was a bit convoluted. There was race condition happening which could cause multiple greenlets to succeed on `let_one_through`

File: exceptions.py [hunk 2]
```diff
@@ -76,24 +76,26 @@ def _handle_exception(self, main_client, **kwargs):
...
-                    main_client._purge_client(self.region_client)
-                    # Sleep for an arbitrary amount of time. If this returns
-                    # False then we've hit our max retry threshold. Die.
-                    key = self.region_client.host + ':' + self.region_client.port
-                    if not _dynamic_sleep(self, key):
-                        raise self
-            # Notify all the other threads to wake up because we've handled the
-            # exception for everyone!
-            _let_all_through(self, self.region_client)
+            try:
```
Comment (lzheng) [inline exceptions.py:79]: Should it be:
try: 
   if _let_one_through(self, self.region_client):
   ....
finally:
   _let_all_through()?

It seems _let_one_through will always get my_sem no matter it returns true or false.

Comment (spicavigo) [inline exceptions.py:79]: No.

Only when it returns True, it does not release the semaphore. When it returns False, it has released the semaphore already

**Comment by lzheng** [comment]
@spicavigo It looks good to me

**Pull request merged**